src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -6499,15 +6517,105 @@ int RGWRados::set_attrs(const DoutPrefixProvider *dpp, RGWObjectCtx* rctx, RGWBu
   return 0;
 }
 
+static int get_part_obj_state(const DoutPrefixProvider* dpp, optional_yield y,
+                              RGWRados* store, RGWBucketInfo& bucket_info,
+                              RGWObjectCtx* rctx, RGWObjManifest* manifest,
+                              int part_num, int* parts_count, bool prefetch,
+                              RGWObjState** pstate, RGWObjManifest** pmanifest)
+{
+  // navigate to the requested part in the manifest
+  RGWObjManifest::obj_iterator end = manifest->obj_end(dpp);
+  if (end.get_cur_part_id() == 0) { // not multipart
+    ldpp_dout(dpp, 20) << "object does not have a multipart manifest" << dendl;
+    return -ERR_INVALID_PART;
+  }
```

Comment:
interesting, a java test case `ObjectTest.testDownloadHLAPI` is failing on this because it's requesting `partNumber=1` for a non-multipart upload. i guess it expects to read the entire object in that case?

i suppose that could be used as an optimization to read `partNumber=1` which, if multipart, will return the `x-amz-mp-parts-count` that you can then read as parts in parallel

Reply:
updated the `test_non_multipart_get_part` test case for this and see it passing